source/s3_client.c

```diff
@@ -668,13 +668,9 @@ int s_apply_endpoint_override(
     }
 
     struct aws_byte_cursor host_value;
-    if (aws_http_headers_get(message_headers, g_host_header_name, &host_value)) {
-        AWS_LOGF_ERROR(
-            AWS_LS_S3_CLIENT,
-            "id=%p Cannot create meta s3 request; message provided in options does not have a 'Host' header.",
-            (void *)client);
-        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
-    }
+    int error = aws_http_headers_get(message_headers, g_host_header_name, &host_value);
+    AWS_ASSERT(!error);
+    (void)error;
```

Comment:
```suggestion
    AWS_ASSERT(aws_http_headers_get(message_headers, g_host_header_name, &host_value));
```

Reply:
AWS_ASSERT will be optimized out for release build.

Reply:
What about precondition if we are asserting that this will never fail?

Reply:
If anything went wrong, we can catch the error during our tests and CI, which basically are debug build. But, anything went wrong from there, will be a programming bug, instead of something can really happen. 

But, yeah, in theory, that should never fail. Just doing this to respect the return code.

Reply:
A fatal assert might be better. We do the same in Swift. If something slips through our tests, this will be a hard to debug bug because of this weird ignoring the error in production (which will cause failure somewhere else) but trying to catch it in debug.

Reply:
The logic here is host header has already been set by the code [above](https://github.com/awslabs/aws-c-s3/pull/255/files#diff-5ee13dd0ee005b828bde6d60ff89582070983e3cb7e5d92389b629a9850c27f8R651-R667). 

It should never ever error out.